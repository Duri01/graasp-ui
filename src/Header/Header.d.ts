import React from 'react'

interface HeaderProps {
  classes: {
    appBar: string
    appBarShift: string
    menuButton: string
    hide: string
  }
  theme: {
    direction: string
  }
  isSidebarOpen: boolean
  handleDrawerOpen: () => {}
}

declare const Header: React.FC<HeaderProps>
