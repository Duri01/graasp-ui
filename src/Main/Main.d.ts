import React from 'react';

interface MainProps {
  classes: {
    fullScreen: string;
    root: string;
    content: string;
    contentShift: string;
    drawerHeader: string;
  };
  fullScreen: boolean;
  children?: React.ReactElement;
  sidebar?: React.ReactElement;
  open?: boolean;
}

declare const Main: React.Component<MainProps>;
