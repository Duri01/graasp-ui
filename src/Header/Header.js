import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import clsx from 'clsx'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import IconButton from '@material-ui/core/IconButton'
import MenuIcon from '@material-ui/icons/Menu'
import { DRAWER_WIDTH } from '../constants'

const styles = (theme) => ({
  appBar: {
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    })
  },
  appBarShift: {
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    marginLeft: DRAWER_WIDTH,
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  menuButton: {
    marginLeft: theme.spacing(2),
    marginRight: theme.spacing(3)
  },
  hide: {
    display: 'none'
  }
})

const Header = ({ id, classes, handleDrawerOpen, isSidebarOpen }) => {
  return (
    <AppBar
      position='fixed'
      className={clsx(classes.appBar, {
        [classes.appBarShift]: isSidebarOpen
      })}
    >
      <Toolbar disableGutters={!isSidebarOpen}>
        <IconButton
          id={id}
          color='inherit'
          aria-label='Open drawer'
          onClick={handleDrawerOpen}
          className={clsx(classes.menuButton, isSidebarOpen && classes.hide)}
        >
          <MenuIcon />
        </IconButton>
      </Toolbar>
    </AppBar>
  )
}

const StyledComponent = withStyles(styles, { withTheme: true })(Header)

export default StyledComponent
