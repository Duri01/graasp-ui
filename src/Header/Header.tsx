import React, { FC } from 'react';
import { makeStyles } from '@mui/styles';
import clsx from 'clsx';
import AppBar from '@mui/material/AppBar';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import { DRAWER_WIDTH } from '../constants';
import { OPEN_DRAWER_LABEL } from '../texts';

export type HeaderProps = {
  id?: string;
  isSidebarOpen?: boolean;
  handleDrawerOpen?: () => void;
  hasSidebar: boolean;
  openDrawerAriaLabel?: string;
  leftContent?: React.ReactElement;
  rightContent?: React.ReactElement;
};

const useStyles = makeStyles((theme) => ({
  appBar: {
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    marginLeft: DRAWER_WIDTH,
    transition: theme.transitions.create(['margin', 'width'], {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  menuButton: {
    marginLeft: theme.spacing(2),
    marginRight: theme.spacing(3),
  },
  hide: {
    display: 'none',
  },
  toolbar: {
    justifyContent: 'space-between',
  },
}));

export const Header: FC<HeaderProps> = ({
  id,
  hasSidebar,
  handleDrawerOpen,
  isSidebarOpen = false,
  openDrawerAriaLabel = OPEN_DRAWER_LABEL,
  leftContent,
  rightContent,
}) => {
  const classes = useStyles();
  return (
    <AppBar
      position='fixed'
      className={clsx(classes.appBar, {
        [classes.appBarShift]: isSidebarOpen,
      })}
    >
      <Toolbar disableGutters={!isSidebarOpen} className={classes.toolbar}>
        {hasSidebar && (
          <IconButton
            id={id}
            color='inherit'
            aria-label={openDrawerAriaLabel}
            onClick={handleDrawerOpen}
            className={clsx(classes.menuButton, {
              [classes.hide]: isSidebarOpen,
            })}
          >
            <MenuIcon />
          </IconButton>
        )}
        {leftContent}
        {rightContent}
      </Toolbar>
    </AppBar>
  );
};

export default Header;
