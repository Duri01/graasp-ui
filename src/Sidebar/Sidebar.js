import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import Drawer from '@material-ui/core/Drawer'
import DrawerHeader from '../DrawerHeader'
import { DRAWER_WIDTH } from '../constants'

const styles = () => ({
  drawer: {
    width: DRAWER_WIDTH,
    flexShrink: 0
  },
  drawerPaper: {
    width: DRAWER_WIDTH
  }
})

export const Sidebar = ({
  className,
  classes,
  isSidebarOpen = false,
  handleDrawerClose,
  children
}) => {
  return (
    <Drawer
      className={className}
      variant='persistent'
      anchor='left'
      open={isSidebarOpen}
      classes={{
        paper: classes.drawerPaper
      }}
    >
      <DrawerHeader handleDrawerClose={handleDrawerClose} />
      {children}
    </Drawer>
  )
}

const StyledComponent = withStyles(styles, { withTheme: true })(Sidebar)

export default StyledComponent
