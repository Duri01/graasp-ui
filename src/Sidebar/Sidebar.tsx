import React, { FC } from 'react';
import Drawer from '@mui/material/Drawer';
import DrawerHeader from '../DrawerHeader';
import { DRAWER_WIDTH } from '../constants';
import { styled, SxProps } from '@mui/material';

export interface SidebarProps {
  sx?: SxProps;
  isSidebarOpen?: boolean;
  handleDrawerClose?: () => void;
  children?: React.ReactElement;
  drawerHeaderContent?: React.ReactElement;
}

export const Sidebar: FC<SidebarProps> = ({
  sx,
  isSidebarOpen = false,
  handleDrawerClose,
  children,
  drawerHeaderContent,
}) => {
  const SideBar = styled(Drawer)({
    width: DRAWER_WIDTH,
    flexShrink: 0,
    '.MuiDrawer-paper': {
      width: DRAWER_WIDTH,
    },
  });
  return (
    <SideBar sx={sx} variant='persistent' anchor='left' open={isSidebarOpen}>
      <DrawerHeader handleDrawerClose={handleDrawerClose}>
        {drawerHeaderContent}
      </DrawerHeader>
      {children}
    </SideBar>
  );
};

export default Sidebar;
