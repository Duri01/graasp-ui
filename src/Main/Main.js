import React, { Component } from 'react';
import clsx from 'clsx';
import { withStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import Header from '../Header';
import Sidebar from '../Sidebar';
import { DRAWER_WIDTH } from '../constants';

const styles = (theme) => ({
  root: {
    display: 'flex',
    height: '100%',
  },
  fullScreen: {
    width: '100%',
    height: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  content: {
    flexGrow: 1,
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  contentShift: {
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginLeft: DRAWER_WIDTH,
  },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    padding: '0 8px',
    ...theme.mixins.toolbar,
    justifyContent: 'flex-end',
  },
});

export class Main extends Component {
  state = (() => ({
    open: this.props.open || false,
  }))();

  static defaultProps = {
    fullScreen: false,
  };

  handleDrawerOpen = () => {
    this.setState({ open: true });
  };

  handleDrawerClose = () => {
    this.setState({ open: false });
  };

  render() {
    const {
      classes,
      children,
      fullScreen,
      sidebar,
      headerLeftContent,
      headerRightContent,
    } = this.props;
    const { open } = this.state;
    const hasSidebar = Boolean(sidebar);

    return (
      <div className={classes.root}>
        <CssBaseline />
        <Header
          hasSidebar={hasSidebar}
          isSidebarOpen={open}
          handleDrawerOpen={this.handleDrawerOpen}
          leftContent={headerLeftContent}
          rightContent={headerRightContent}
        />

        {hasSidebar && (
          <Sidebar
            isSidebarOpen={open}
            handleDrawerClose={this.handleDrawerClose}
          >
            {sidebar}
          </Sidebar>
        )}

        <main
          className={clsx(classes.content, {
            [classes.contentShift]: open,
            [classes.fullScreen]: fullScreen,
          })}
        >
          <div className={classes.drawerHeader} />
          {children}
        </main>
      </div>
    );
  }
}

Main.defaultProps = {
  headerLeftContent: null,
  headerRightContent: null,
};

const StyledComponent = withStyles(styles, { withTheme: true })(Main);

export default StyledComponent;
