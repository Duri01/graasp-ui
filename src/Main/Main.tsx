import React, { Component } from 'react';
import clsx from 'clsx';
import {
  withStyles,
  Theme,
  createStyles,
  WithStyles,
} from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import Header from '../Header';
import Sidebar from '../Sidebar';
import { DRAWER_WIDTH } from '../constants';

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const styles = (theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      height: '100%',
    },
    fullScreen: {
      width: '100%',
      height: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    content: {
      flexGrow: 1,
      transition: theme.transitions.create('margin', {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen,
      }),
    },
    contentShift: {
      transition: theme.transitions.create('margin', {
        easing: theme.transitions.easing.easeOut,
        duration: theme.transitions.duration.enteringScreen,
      }),
      marginLeft: DRAWER_WIDTH,
    },
    drawerHeader: {
      display: 'flex',
      alignItems: 'center',
      padding: '0 8px',
      ...theme.mixins.toolbar,
      justifyContent: 'flex-end',
    },
  });

export interface MainProps extends WithStyles<typeof styles> {
  fullScreen?: boolean;
  children?: React.ReactElement;
  sidebar?: React.ReactElement;
  open?: boolean;
  headerLeftContent?: React.ReactElement;
  headerRightContent?: React.ReactElement;
}

type MainState = {
  open: boolean;
};

export class Main extends Component<MainProps, MainState> {
  state = ((): MainState => ({
    open: this.props.open || false,
  }))();

  static defaultProps = {
    fullScreen: false,
  };

  handleDrawerOpen = (): void => {
    this.setState({ open: true });
  };

  handleDrawerClose = (): void => {
    this.setState({ open: false });
  };

  render(): JSX.Element {
    const {
      classes,
      children,
      fullScreen,
      sidebar,
      headerLeftContent,
      headerRightContent,
    } = this.props;
    const { open } = this.state;
    const hasSidebar = Boolean(sidebar);

    return (
      <div className={classes.root}>
        <CssBaseline />
        <Header
          hasSidebar={hasSidebar}
          isSidebarOpen={open}
          handleDrawerOpen={this.handleDrawerOpen}
          leftContent={headerLeftContent}
          rightContent={headerRightContent}
        />

        {hasSidebar && (
          <Sidebar
            isSidebarOpen={open}
            handleDrawerClose={this.handleDrawerClose}
          >
            {sidebar}
          </Sidebar>
        )}

        <main
          className={clsx(classes.content, {
            [classes.contentShift]: open,
            [classes.fullScreen]: fullScreen,
          })}
        >
          <div className={classes.drawerHeader} />
          {children}
        </main>
      </div>
    );
  }
}

export default withStyles(styles, { withTheme: true })(Main);
