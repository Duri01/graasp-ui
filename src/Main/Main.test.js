import React from 'react';
import { shallow } from 'enzyme';
import { Main } from './Main';
import Header from '../Header';
import Sidebar from '../Sidebar';

const createMainProps = ({ fullScreen, children, sidebar } = {}) => ({
  classes: {
    fullScreen: 'fullScreen',
    root: '',
    content: '',
    contentShift: '',
    drawerHeader: '',
  },
  children,
  sidebar,
  fullScreen,
});

describe('Main', () => {
  let wrapper;

  describe('default values', () => {
    beforeAll(() => {
      const props = createMainProps();
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Main {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <Header /> component with default values', () => {
      const sidebar = wrapper.find(Sidebar);
      expect(sidebar).toHaveLength(0);
      const header = wrapper.find(Header);
      expect(header).toHaveLength(1);
      const main = wrapper.find('main');
      expect(main).toHaveLength(1);
      expect(main.hasClass('fullScreen')).toEqual(false);
    });
  });

  describe('fullScreen = true', () => {
    beforeAll(() => {
      const props = createMainProps({ fullScreen: true });
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Main {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <Header /> and no <Sidebar/> components', () => {
      const sidebar = wrapper.find(Sidebar);
      expect(sidebar).toHaveLength(0);
      const header = wrapper.find(Header);
      expect(header).toHaveLength(1);
      const main = wrapper.find('main');
      expect(main).toHaveLength(1);
      expect(main.hasClass('fullScreen')).toEqual(true);
    });
  });

  describe('children is defined', () => {
    beforeAll(() => {
      const props = createMainProps({ children: <child>some child</child> });
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Main {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <Header /> component', () => {
      const sidebar = wrapper.find(Sidebar);
      expect(sidebar).toHaveLength(0);
      const header = wrapper.find(Header);
      expect(header).toHaveLength(1);
      const main = wrapper.find('main');
      expect(main).toHaveLength(1);
      expect(main.hasClass('fullScreen')).toEqual(false);
      const child = wrapper.find('child');
      expect(child).toHaveLength(1);
    });
  });

  describe('sidebar is defined', () => {
    beforeAll(() => {
      const props = createMainProps({ sidebar: <child>some child</child> });
      // eslint-disable-next-line react/jsx-props-no-spreading
      wrapper = shallow(<Main {...props} />);
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders one <Header /> and one <Sidebar/> components', () => {
      const sidebar = wrapper.find(Sidebar);
      expect(sidebar).toHaveLength(1);
      const header = wrapper.find(Header);
      expect(header).toHaveLength(1);
      const main = wrapper.find('main');
      expect(main).toHaveLength(1);
      expect(main.hasClass('fullScreen')).toEqual(false);
      const child = wrapper.find('child');
      expect(child).toHaveLength(1);
    });
  });
});
