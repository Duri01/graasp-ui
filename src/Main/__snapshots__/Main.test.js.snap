// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main children is defined renders correctly 1`] = `
<div
  className=""
>
  <WithStyles(CssBaseline) />
  <WithStyles(Header)
    handleDrawerOpen={[Function]}
    hasSidebar={false}
    isSidebarOpen={false}
    leftContent={null}
    rightContent={null}
  />
  <main
    className=""
  >
    <div
      className=""
    />
    <child>
      some child
    </child>
  </main>
</div>
`;

exports[`Main default values renders correctly 1`] = `
<div
  className=""
>
  <WithStyles(CssBaseline) />
  <WithStyles(Header)
    handleDrawerOpen={[Function]}
    hasSidebar={false}
    isSidebarOpen={false}
    leftContent={null}
    rightContent={null}
  />
  <main
    className=""
  >
    <div
      className=""
    />
  </main>
</div>
`;

exports[`Main fullScreen = true renders correctly 1`] = `
<div
  className=""
>
  <WithStyles(CssBaseline) />
  <WithStyles(Header)
    handleDrawerOpen={[Function]}
    hasSidebar={false}
    isSidebarOpen={false}
    leftContent={null}
    rightContent={null}
  />
  <main
    className="fullScreen"
  >
    <div
      className=""
    />
  </main>
</div>
`;

exports[`Main sidebar is defined renders correctly 1`] = `
<div
  className=""
>
  <WithStyles(CssBaseline) />
  <WithStyles(Header)
    handleDrawerOpen={[Function]}
    hasSidebar={true}
    isSidebarOpen={false}
    leftContent={null}
    rightContent={null}
  />
  <WithStyles(Sidebar)
    handleDrawerClose={[Function]}
    isSidebarOpen={false}
  >
    <child>
      some child
    </child>
  </WithStyles(Sidebar)>
  <main
    className=""
  >
    <div
      className=""
    />
  </main>
</div>
`;
